File: app/src/main/java/com/bis5/fitjourney/fragments/HomeFragment.java
package com.bis5.fitjourney.fragments;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import com.bis5.fitjourney.R;
import com.bis5.fitjourney.databinding.FragmentHomeBinding;
import com.bis5.fitjourney.viewmodels.SharedViewModel;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.List;

public class HomeFragment extends Fragment {

    private FragmentHomeBinding binding;
    private SharedViewModel sharedViewModel;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentHomeBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        sharedViewModel = new ViewModelProvider(requireActivity()).get(SharedViewModel.class);

        setupNavigation();
        setupWelcomeMessage();
        setupQuoteOfTheDay();
        setupBarChart();
    }

    private void setupWelcomeMessage() {
        sharedViewModel.getUserEmail().observe(getViewLifecycleOwner(), email -> {
            if (email != null && !email.isEmpty()) {
                binding.tvWelcome.setText(String.format("Welcome, %s!", email.split("@")[0]));
            }
        });
    }

    private void setupQuoteOfTheDay() {
        List<String> quotes = Arrays.asList(
                "'The body achieves what the mind believes.'",
                "'The only bad workout is the one that didn't happen.'",
                "'Success isn't always about greatness. It's about consistency. Consistent hard work gains success. Greatness will come.'",
                "'The clock is ticking. Are you becoming the person you want to be?'",
                "'Your body can stand almost anything. It's your mind that you have to convince.'"
        );
        int dayOfYear = Calendar.getInstance().get(Calendar.DAY_OF_YEAR);
        binding.tvQuoteOfTheDay.setText(quotes.get(dayOfYear % quotes.size()));
    }

    private void setupBarChart() {
        List<BarEntry> entries = new ArrayList<>();
        entries.add(new BarEntry(0, 4f));
        entries.add(new BarEntry(1, 8f));
        entries.add(new BarEntry(2, 6f));
        entries.add(new BarEntry(3, 2f));
        entries.add(new BarEntry(4, 10f));
        entries.add(new BarEntry(5, 5f));
        entries.add(new BarEntry(6, 7f));

        BarDataSet dataSet = new BarDataSet(entries, "Weekly Activity");
        int chartColor = ContextCompat.getColor(requireContext(), R.color.a_500);
        dataSet.setColor(chartColor);
        dataSet.setValueTextColor(Color.WHITE);

        BarData barData = new BarData(dataSet);
        binding.barChart.setData(barData);

        // Customize the chart
        binding.barChart.getDescription().setEnabled(false);
        binding.barChart.getLegend().setEnabled(false);
        binding.barChart.setDrawValueAboveBar(false);
        binding.barChart.setFitBars(true);

        // Customize X-axis
        XAxis xAxis = binding.barChart.getXAxis();
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setDrawGridLines(false);
        xAxis.setTextColor(Color.WHITE);
        final String[] days = {"Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"};
        xAxis.setValueFormatter(new IndexAxisValueFormatter(days));

        // Customize Y-axis
        binding.barChart.getAxisLeft().setTextColor(Color.WHITE);
        binding.barChart.getAxisRight().setEnabled(false);

        binding.barChart.invalidate(); // Refresh the chart
    }

    private void setupNavigation() {
        binding.cardWorkouts.setOnClickListener(v -> 
            NavHostFragment.findNavController(this).navigate(R.id.action_homeFragment_to_workoutListFragment));

        binding.cardNutrition.setOnClickListener(v -> 
            NavHostFragment.findNavController(this).navigate(R.id.action_homeFragment_to_nutritionLogFragment));
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/bis5/fitjourney/fragments/HomeFragment.kt
package com.bis5.fitjourney.fragments

import android.app.AlertDialog
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import com.bis5.fitjourney.R
import com.bis5.fitjourney.databinding.FragmentHomeBinding
import com.bis5.fitjourney.models.AppDatabase
import com.bis5.fitjourney.models.SetLog
import com.bis5.fitjourney.models.Workout
import com.bis5.fitjourney.viewmodels.WorkoutViewModel
import com.bis5.fitjourney.viewmodels.WorkoutViewModelFactory
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import java.util.Calendar
import java.util.concurrent.TimeUnit

class HomeFragment : Fragment() {

    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!

    private lateinit var workoutViewModel: WorkoutViewModel
    private var userEmail: String? = null
    private var selectedChartExercise = "Bench Press" // Default

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        userEmail = requireActivity().intent.getStringExtra("USER_EMAIL")
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val database = AppDatabase.getDatabase(requireContext())
        val viewModelFactory = WorkoutViewModelFactory(database.workoutDao(), database.exerciseDao(), database.workoutLogDao(), database.setLogDao())
        workoutViewModel = ViewModelProvider(this, viewModelFactory)[WorkoutViewModel::class.java]

        userEmail?.let { email ->
            workoutViewModel.createDefaultWorkoutsIfNoneExist(email) // This is the new line

            binding.tvUserName.text = email.split("@").firstOrNull()?.replaceFirstChar(Char::titlecase) ?: "User"
            observeNextWorkout(email)
            observeWorkoutLogs(email)
            observeChartData(email)
        }

        binding.btnGoToWorkout.setOnClickListener {
            findNavController().navigate(HomeFragmentDirections.actionHomeFragmentToWorkoutListFragment())
        }

        binding.ivSettings.setOnClickListener {
            findNavController().navigate(HomeFragmentDirections.actionHomeFragmentToProfileFragment())
        }

        binding.ivChartOptions.setOnClickListener {
            showExerciseSelectionDialog()
        }
    }

    private fun observeNextWorkout(email: String) {
        workoutViewModel.getWorkouts(email).observe(viewLifecycleOwner) { workouts ->
            if (workouts.isNotEmpty()) {
                val nextWorkout = workouts.maxByOrNull { it.date }!!
                binding.tvNextWorkoutName.text = nextWorkout.name
                updateNextWorkoutIcon(nextWorkout)
                binding.cardNextWorkout.setOnClickListener {
                    findNavController().navigate(HomeFragmentDirections.actionHomeFragmentToWorkoutDetailFragment(nextWorkout.id))
                }
            } else {
                binding.tvNextWorkoutName.text = "No workouts yet!"
                binding.ivNextWorkoutIcon.visibility = View.GONE
            }
        }
    }

    private fun observeWorkoutLogs(email: String) {
        workoutViewModel.getAllLogsForUser(email).observe(viewLifecycleOwner) { logs ->
            updateHeatmap(logs.map { it.dateStarted })
        }
    }

    private fun observeChartData(email: String) {
        workoutViewModel.getSetHistoryForExercise(email, selectedChartExercise).observe(viewLifecycleOwner) { setLogs ->
            if (setLogs.isNotEmpty()) {
                setupProgressChart(setLogs)
            }
        }
    }

    private fun showExerciseSelectionDialog() {
        userEmail?.let { email ->
            workoutViewModel.getUniqueExerciseNames(email).observe(viewLifecycleOwner) { exerciseNames ->
                val builder = AlertDialog.Builder(requireContext())
                builder.setTitle("Select Exercise for Chart")
                builder.setItems(exerciseNames.toTypedArray()) { _, which ->
                    selectedChartExercise = exerciseNames[which]
                    binding.tvChartTitle.text = "Progress: $selectedChartExercise"
                    observeChartData(email) // Re-observe with the new exercise
                }
                builder.show()
            }
        }
    }

    private fun updateNextWorkoutIcon(workout: Workout) {
        val iconRes = when (workout.name.lowercase()) {
            "push day" -> R.drawable.push
            "pull day" -> R.drawable.pull
            "leg day" -> R.drawable.legs
            "upper body" -> R.drawable.upper
            "lower body" -> R.drawable.lower
            "core" -> R.drawable.core
            "cardio" -> R.drawable.cardio
            else -> R.drawable.dumbell
        }
        binding.ivNextWorkoutIcon.setImageResource(iconRes)
    }

    private fun setupProgressChart(setLogs: List<SetLog>) {
        val entries = ArrayList<Entry>()
        if (setLogs.isEmpty()) return

        val referenceDate = setLogs.first().timestamp
        setLogs.forEach { log ->
            val daysSinceStart = TimeUnit.MILLISECONDS.toDays(log.timestamp - referenceDate).toFloat()
            entries.add(Entry(daysSinceStart, log.weight.toFloat()))
        }

        val dataSet = LineDataSet(entries, "Weight (kg)")
        val lineData = LineData(dataSet)

        binding.progressChart.data = lineData
        binding.progressChart.description.isEnabled = false
        binding.progressChart.invalidate() // refresh
    }

    private fun updateHeatmap(workoutDates: List<Long>) {
        binding.heatmapContainer.removeAllViews()
        val cal = Calendar.getInstance()

        for (i in 6 downTo 0) {
            cal.timeInMillis = System.currentTimeMillis()
            cal.add(Calendar.DAY_OF_YEAR, -i)
            val dayView = View(context)
            val size = resources.getDimensionPixelSize(R.dimen.heatmap_day_size)
            val margin = resources.getDimensionPixelSize(R.dimen.heatmap_day_margin)
            val params = LinearLayout.LayoutParams(size, size)
            params.setMargins(margin, 0, margin, 0)
            dayView.layoutParams = params

            val workoutOnThisDay = workoutDates.any { date ->
                val workoutCal = Calendar.getInstance()
                workoutCal.timeInMillis = date
                workoutCal.get(Calendar.YEAR) == cal.get(Calendar.YEAR) && workoutCal.get(Calendar.DAY_OF_YEAR) == cal.get(Calendar.DAY_OF_YEAR)
            }

            val circleDrawable = ContextCompat.getDrawable(requireContext(), R.drawable.heatmap_day_circle)?.mutate() as GradientDrawable
            val colorRes = if (workoutOnThisDay) R.color.md_theme_primary else R.color.md_theme_surfaceVariant
            circleDrawable.setColor(ContextCompat.getColor(requireContext(), colorRes))
            dayView.background = circleDrawable

            binding.heatmapContainer.addView(dayView)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
